// Copyright: Ankitects Pty Ltd and contributors
// License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html

syntax = "proto3";

option java_multiple_files = true;

package anki.revlog;

import "anki/generic.proto";
import "anki/collection.proto";

service RevlogService {
  rpc GetRevlogEntry(RevlogId) returns (RevlogEntry);
  rpc UpdateRevlogEntries(UpdateRevlogEntriesRequest)
      returns (collection.OpChanges);
  rpc RemoveRevlogEntries(RemoveRevlogEntriesRequest)
      returns (collection.OpChangesWithCount);
}

message RevlogId {
  int64 rid = 1;
}

message RevlogIds {
  repeated int64 rids = 1;
}

// RevlogEntry was moved here from proto/anki/stats.proto.

message RevlogEntry {
  enum ReviewKind {
    LEARNING = 0;
    REVIEW = 1;
    RELEARNING = 2;
    // Recent Anki versions only use this when rescheduling disabled
    FILTERED = 3;
    MANUAL = 4;
  }
  int64 id = 1;
  int64 cid = 2;
  int32 usn = 3;
  uint32 button_chosen = 4;
  int32 interval = 5;
  int32 last_interval = 6;
  uint32 ease_factor = 7;
  uint32 taken_millis = 8;
  ReviewKind review_kind = 9;

  // The following three fields were added to support a version of Anki for
  // detailed review feedback.
  uint32 mtime_secs = 10;
  string feedback = 11;
  repeated string tags = 12;
}

message UpdateRevlogEntriesRequest {
  repeated RevlogEntry revlog_entries = 1;
  bool skip_undo_entry = 2;
}

message RemoveRevlogEntriesRequest {
  repeated int64 revlog_ids = 1;
}
